### PR TITLE
Avoid returning async void AB#404455

### PR DESCRIPTION
**Pull request by pankti11**

### CODE REVIEW

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		private static DiagnosticDescriptor Rule = new DiagnosticDescriptor(Helper.ToDiagnosticId(DiagnosticIds.AvoidTaskVoid), Title, MessageFormat, Category, DiagnosticSeverity.Error, isEnabledByDefault: true, description: Description, helpLinkUri: HelpUri);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+		public string Identifier = @"void";
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:34]: I'd propose not running it on generated code.

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+			{
+				if (startContext.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Task") == null)
+				{
+					return;
+				}
+				startContext.RegisterSyntaxNodeAction(AnalyzeIfStatement, SyntaxKind.MethodDeclaration);
+			});
+		}
+
+		private void AnalyzeIfStatement(SyntaxNodeAnalysisContext context)
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:45]: AnalyzeIfStatement?

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		{
+			INamedTypeSymbol taskSymbol = context.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Task");
+			MethodDeclarationSyntax methodDeclaration = (MethodDeclarationSyntax)context.Node;
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:48]: We shouldn't load the semantic model for every method - for performance reasons.  We can/should filter on the async void syntax first (and potentially not even needing to load the semantic model, unless it's needed for event handlers).

**Review by bcollamore — CHANGES_REQUESTED** [review]

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+using Philips.CodeAnalysis.Common;
+
+namespace Philips.CodeAnalysis.MaintainabilityAnalyzers.Maintainability
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public class AvoidAsyncVoidAnalyzer : DiagnosticAnalyzer
+	{
+		private const string Title = @"Avoid async void";
+		public const string MessageFormat = @"Methods may not have async void return type";
+		private const string Description = @"To avoid unhandled exception, methods should use async void unless a event handler.";
```
[the remark this answers is not in the document]
Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs]: done

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs (hunk 1 above)
Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs]: changed it to None

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs (hunk 2 above)
Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs]: changed it to Analyze

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs (hunk 3 above)
Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:45]: changed the logic to filter out some methods

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+				startContext.RegisterSyntaxNodeAction(Analyze, SyntaxKind.MethodDeclaration);
+			});
+		}
+
+		private void Analyze(SyntaxNodeAnalysisContext context)
+		{
+			MethodDeclarationSyntax methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+
+			if (!methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword))
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:48]: I think MethodDeclarationSyntax also has a ReturnType, for which you can compare with void.  Ie if not void, return.

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			}
+
+			ISymbol symbol = context.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+			if (symbol is null || !(symbol is IMethodSymbol methodSymbol))
+			{
+				return;
+			}
+
+			INamedTypeSymbol taskSymbol = context.Compilation.GetTypeByMetadataName("System.Threading.Tasks.Task");
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:60]: It's not clear to me why we need this check.  Ie., if it's async void then it's a problem...

**Review by bcollamore — CHANGES_REQUESTED** [review]

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs (hunk 5 above)
Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs]: Yes. I added that check with a void return type. I think made it little complicated before. Now the logic is simple and better.

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs (hunk 6 above)
Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs]: @bcollamore  yes removed this complicated check

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 7]
```diff
@@ -45,7 +45,8 @@ private void Analyze(SyntaxNodeAnalysisContext context)
 			MethodDeclarationSyntax methodDeclaration = (MethodDeclarationSyntax)context.Node;
 
 
-			if (!methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword))
+			if (!methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword) ||
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:48]: You don't need to hardcode "void".  There's a way to compare with SyntaxKind.VoidKeyword.

**Review by bcollamore — CHANGES_REQUESTED** [review]

File: Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		{
+			MethodDeclarationSyntax methodDeclaration = (MethodDeclarationSyntax)context.Node;
+
+
+			if (!methodDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword))
+			{
+				return;
+			}
+
+			if ((methodDeclaration.ReturnType is not PredefinedTypeSyntax predefinedTypeSyntax) || predefinedTypeSyntax.Keyword.Kind() != SyntaxKind.VoidKeyword)
```
Comment (bcollamore) [inline Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs:52]: See this?  The feature 'not pattern' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.

Comment (pankti11) [reply Philips.CodeAnalysis.MaintainabilityAnalyzers/Maintainability/AvoidAsyncVoidAnalyzer.cs]: Changed it.

**Review by bcollamore — APPROVED** [review]

**Pull request merged**